protocol/contracts/TrucoMatch.sol

```diff
@@ -10,42 +10,131 @@ import "./crypt/DeckCrypt.sol";
+    event PlayerStaked(address player, uint256 tokensAtStake);
+    event DealStarted(address shuffler);
+    event DealEnded();
+
+
+    constructor (IERC3333 _trucoEngine, DeckCrypt _deckCrypt, IERC20 _truCoin, uint256 _tokensAtStake) {
 
-    constructor (IERC3333 _trucoEngine, DeckCrypt _deckCrypt, IERC20 _trucoin) {
         trucoEngine = _trucoEngine;
         deckCrypt = _deckCrypt;
-        currentMatch.players[0] = msg.sender;
+        truCoin = _truCoin;
+        currentMatch.tokensAtStake = _tokensAtStake;
+        currentMatch.players[0].playerAddress = msg.sender;
+
+        //Mint Sould Bound Token
 
-        //Transfer Trucoins from owner to contract
-        //Mint Sould Bound Token 
+        emit MatchCreated(address(this));
     }
-    
         
-    // Method for joining the match 
+    // Method for second player joining the match 
     function join() public {
-        // Check match state
+        require(currentMatch.players[0].playerAddress != msg.sender, "Match creator is already joined");
+        require(currentMatch.players[1].playerAddress == address(0), "Match is full");
+
         // Check invitations if any
-        // Transfer tokens
-        // Close game for new joins
-        // Emit an event
+
+        currentMatch.players[1].playerAddress = msg.sender;
+
+        // Make the user stake so he can not kidnap the match
+        stake(1);
+    }
+
+    // Method for staking & starting the match
+    function stake(uint8 _player) public {
+        require(
+            currentMatch.players[_player].playerAddress == msg.sender, 
+            "Player is not joined");
+        require(
+            truCoin.balanceOf(msg.sender) >= currentMatch.tokensAtStake, 
+            "You don't have enough Trucoins to stake this match");
+        require(
+            truCoin.allowance(msg.sender, address(this)) >= currentMatch.tokensAtStake, 
+            "Not enough trucoins transfer approved");
+        require(
+            currentMatch.players[_player].tokensAtStake == 0, 
+            "You already staked this match");
+
+        // Transfer Trucoins from player to contract
+        require(truCoin.transferFrom(
+            msg.sender, address(this), currentMatch.tokensAtStake), 
+            "Trucoin transfer failed");
+        
+        currentMatch.players[_player].tokensAtStake = currentMatch.tokensAtStake;
+
+        // emit event
+        emit PlayerStaked(msg.sender, currentMatch.tokensAtStake);
+
+        // Start match
+        if (currentMatch.players[0].tokensAtStake == currentMatch.tokensAtStake && 
+            currentMatch.players[1].tokensAtStake == currentMatch.tokensAtStake) {
+            currentMatch.gameState = trucoEngine.startGame();
+            emit MatchStarted(
+                currentMatch.players[0].playerAddress, 
+                currentMatch.players[1].playerAddress,
+                currentMatch.tokensAtStake);
+        }
```

Comment:
At this point `Engine.startGame` should be called, where it deducts fees and returns the initial game state to work with